packages/uploadthing/src/h3.ts

```diff
@@ -0,0 +1,81 @@
+import type { H3Event } from "h3";
+import {
+  createRouter,
+  defineEventHandler,
+  getRequestHeaders,
+  getRequestURL,
+  readBody,
+  setHeaders,
+  setResponseStatus,
+} from "h3";
+
+import type { Json } from "@uploadthing/shared";
+import { getStatusCodeFromError, UploadThingError } from "@uploadthing/shared";
+
+import { UPLOADTHING_VERSION } from "./constants";
+import { defaultErrorFormatter } from "./internal/error-formatter";
+import {
+  buildPermissionsInfoHandler,
+  buildRequestHandler,
+} from "./internal/handler";
+import type { RouterWithConfig } from "./internal/handler";
+import type { FileRouter } from "./internal/types";
+import type { CreateBuilderOptions } from "./internal/upload-builder";
+import { createBuilder } from "./internal/upload-builder";
+
+export type { FileRouter } from "./internal/types";
+
+export const createUploadthing = <TErrorShape extends Json>(
+  opts?: CreateBuilderOptions<TErrorShape>,
+) =>
+  createBuilder<
+    { req: undefined; res: undefined; event: H3Event },
+    TErrorShape
+  >(opts);
+
+export const createH3EventHandler = <TRouter extends FileRouter>(
+  opts: RouterWithConfig<TRouter>,
+) => {
+  const requestHandler = buildRequestHandler(opts);
+  const POST = defineEventHandler(async (event) => {
+    const response = await requestHandler({
+      req: {
+        url: getRequestURL(event).href,
+        headers: getRequestHeaders(event),
+        json: () => Promise.resolve(readBody(event)),
+      },
+      event,
+    });
+
+    setHeaders(event, { "x-uploadthing-version": UPLOADTHING_VERSION });
+
+    if (response instanceof UploadThingError) {
+      setResponseStatus(event, getStatusCodeFromError(response));
+      const errorFormatter =
+        opts.router[Object.keys(opts.router)[0]]?._def.errorFormatter ??
+        defaultErrorFormatter;
+      return errorFormatter(response) as unknown;
+    }
+
+    if (response.status !== 200) {
+      // We messed up - this should never happen
+      setResponseStatus(event, 500);
+      return "An unknown error occurred";
+    }
+
+    setResponseStatus(event, 200);
+    return response.body;
+  });
+
+  const getBuildPerms = buildPermissionsInfoHandler<TRouter>(opts);
+  const GET = defineEventHandler((event) => {
+    setResponseStatus(event, 200);
+    setHeaders(event, { "x-uploadthing-version": UPLOADTHING_VERSION });
+    return getBuildPerms();
+  });
+
+  const router = createRouter()
+    .post("/api/uploadthing", POST)
+    .get("/api/uploadthing", GET);
+  return router.handler;
```

Comment:
@danielroe what would you return here? we later want to support arbitrary pathnames i.e. not locking in to /api/uploadthing, does that mean I should set the path in here to just / and then let the user put in the /api/uploadthing?

```suggestion
  const router = createRouter()
    .post("/", POST)
    .get("/", GET);
  return router.handler;
```

```ts
// user code
const app = createApp();
app.use(
  "/api/uploadthing", 
  createH3EventHandler({ router: uploadRouter })
);
```

or what would you recommend here that will be adaptable to vanilla h3, nitro, nuxt, solidstart etc etc?

Reply:
Yes, that makes sense to me. We'd then have a Nuxt module, for example, that could have a user option for the path, but default to `/api/uploadthing`. But that way it's also fully configurable if using just Nitro, or another framework.